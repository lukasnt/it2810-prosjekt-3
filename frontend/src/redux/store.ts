import { createStore, combineReducers, Store } from "redux";
import { SearchResult, searchResultReducer } from "./reducers/searchresult";
import { SearchParams, searchParamsReducer } from "./reducers/searchparams";
import { User, userReducer } from "./reducers/user";

export type AppState = {
    user: User | null;
    searchParams: SearchParams;
    searchResult: SearchResult | null;
};

//Utility-funksjon for Ã¥ kombinere flere reducere
const rootReducer = combineReducers<AppState>({
    user: userReducer,
    searchParams : searchParamsReducer,
    searchResult: searchResultReducer
});

function configureStore(): Store<AppState> {
    const store = createStore(rootReducer, undefined);
    return store;
}

//Oppretter en store
export const store = configureStore();