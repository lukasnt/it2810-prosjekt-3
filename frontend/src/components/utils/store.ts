import { createStore, combineReducers, Store } from "redux";

//Typene brukt i state
export type User = {
    email: string;
    firstName: string;
    lastName: string;
    token: string;
};

export type AppState = {
    user: User | null;
};

//Funksjoner som returnerer action-objekter
export function setUser(user : User) {
    return {
        type: "SET_USER",
        payload: user
    } as const;
}
export function removeUser() {
    return {
        type: "REMOVE_USER"
    } as const;
}
type Actions = ReturnType<typeof setUser> | ReturnType<typeof removeUser>;

//Reducer-funksjonen, initialiserer store med user
function userReducer(state : User | null = null, action: Actions) {
    switch (action.type) {
        case "SET_USER":
            return state = ({ 
                    email: action.payload.email, 
                    firstName: action.payload.firstName,
                    lastName: action.payload.lastName,
                    token: action.payload.token 
                });
        case "REMOVE_USER":
            return state = null;
        default:
            neverReached(action);
    }
    return state;
}
function neverReached(never: never) {}

//Utility-funksjon for Ã¥ kombinere flere reducere
const rootReducer = combineReducers<AppState>({
    user: userReducer
});

function configureStore(): Store<AppState> {
    const store = createStore(rootReducer, undefined);
    return store;
}

//Oppretter en store
export const store = configureStore();