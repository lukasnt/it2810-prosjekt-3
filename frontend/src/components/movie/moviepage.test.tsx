import React from 'react';
import { act, render, wait } from '@testing-library/react';
import {Route, MemoryRouter} from 'react-router-dom';
import fetch from 'jest-fetch-mock';
import MoviePage from './moviepage';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from '../../redux/store';


describe('Snapshot test suite for single movie page', () => {

    let component : any = null;

    const {act} = renderer;

    beforeAll(async () => {
        fetch.resetMocks();

        fetch.mockResponseOnce(JSON.stringify({
            primaryTitle: 'Movie',
            startYear: '2020',
            runtimeMinutes: '60',
            genres: 'Action,Comedy',
            posterPath: '../../res/mock_image.jpg',
            voteAverage: '8.6',
            voteCount: '10',
            overview: 'Overview'
        }))

        await act(async () => {
            component = renderer.create(
                <Provider store={store}>
                    <MemoryRouter initialEntries={['/movie/01']}>
                        <Route path='/movie/:tconst' >
                            <MoviePage />
                        </Route>
                    </MemoryRouter>
                </Provider>
            );
        })

        // await wait(() => expect(fetch).toHaveBeenCalled())

    })
    

    it('renders single movie page', () => {
        let tree : any = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
})


describe('Test suite for single movie page', () => {

    let wrapper : any = null;

    beforeEach(async () => {
        fetch.resetMocks();

        fetch.mockResponseOnce(JSON.stringify({
            primaryTitle: 'Movie',
            startYear: '2020',
            runtimeMinutes: '60',
            genres: 'Action,Comedy',
            posterPath: '../../res/mock_image.jpg',
            voteAverage: '8.6',
            voteCount: '10',
            overview: 'Overview'
        }))

        await act(async () => {
            wrapper = render(
            <Provider store={store}>
                <MemoryRouter initialEntries={['/movie/01']}>
                    <Route path='/movie/:tconst' >
                    <MoviePage />
                    </Route>
                </MemoryRouter>
            </Provider>
            )
        })

        // await wait(() => expect(fetch).toHaveBeenCalled())
    });

    afterAll(() => {
        wrapper = null;
    })

  
    it('Expect one call to fetch()', () => {
        expect(fetch).toHaveBeenCalledTimes(1)
    })
    
    it('Test that title and year shows correct data', () => {
        expect(wrapper.getByTestId('movietitle').innerHTML === 'Movie (2020)').toBeTruthy()
    })
    
    it('Test that genres shows correct data', () => {
        expect(wrapper.getByTestId('genres').innerHTML.includes('Action,Comedy')).toBeTruthy()
    })

    it('Test that overview shows correct data', () => {
        expect(wrapper.getByTestId('overview').innerHTML.includes('Overview')).toBeTruthy()
    })
  
})