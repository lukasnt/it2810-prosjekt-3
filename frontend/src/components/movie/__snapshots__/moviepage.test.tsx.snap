// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test suite for single movie page renders single movie page 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  style={
    Object {
      "backgroundImage": "linear-gradient(to right, rgba(44,44,44,1) 15%, rgba(44,44,44,0.7)), url(../../res/mock_image.jpg)",
      "backgroundPositionY": "50%",
      "backgroundSize": "cover",
      "margin": "0px",
      "minHeight": "95vh",
      "width": "100%",
    }
  }
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-7 MuiGrid-grid-md-9"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "flex-start",
      }
    }
  >
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      style={
        Object {
          "backgroundColor": "transparent",
          "boxShadow": "none",
          "marginLeft": "5%",
          "marginRight": "5%",
        }
      }
    >
      <div
        className="MuiCardContent-root"
        style={
          Object {
            "color": "white",
          }
        }
      >
        <h3
          className="MuiTypography-root MuiTypography-h3"
          data-testid="movietitle"
        >
          Movie (2020)
        </h3>
        <h6
          className="MuiTypography-root MuiTypography-subtitle1"
          data-testid="genres"
          style={
            Object {
              "marginTop": "10px",
            }
          }
        >
          Action,Comedy • 1 h 0 m
        </h6>
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
              "marginTop": "10px",
            }
          }
        >
          <span
            aria-label={null}
            className="MuiRating-root"
            onMouseLeave={[Function]}
            onMouseMove={[Function]}
            role={null}
            style={
              Object {
                "marginRight": "3%",
              }
            }
          >
            <label
              className="MuiRating-label"
              htmlFor="hover-feedback-1"
            >
              <span
                className="MuiRating-icon MuiRating-iconEmpty"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </span>
              <span
                className="MuiRating-visuallyhidden"
              >
                1 Star
              </span>
            </label>
            <input
              checked={false}
              className="MuiRating-visuallyhidden"
              disabled={false}
              id="hover-feedback-1"
              name="hover-feedback"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              type="radio"
              value={1}
            />
            <label
              className="MuiRating-label"
              htmlFor="hover-feedback-2"
            >
              <span
                className="MuiRating-icon MuiRating-iconEmpty"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </span>
              <span
                className="MuiRating-visuallyhidden"
              >
                2 Stars
              </span>
            </label>
            <input
              checked={false}
              className="MuiRating-visuallyhidden"
              disabled={false}
              id="hover-feedback-2"
              name="hover-feedback"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              type="radio"
              value={2}
            />
            <label
              className="MuiRating-label"
              htmlFor="hover-feedback-3"
            >
              <span
                className="MuiRating-icon MuiRating-iconEmpty"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </span>
              <span
                className="MuiRating-visuallyhidden"
              >
                3 Stars
              </span>
            </label>
            <input
              checked={false}
              className="MuiRating-visuallyhidden"
              disabled={false}
              id="hover-feedback-3"
              name="hover-feedback"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              type="radio"
              value={3}
            />
            <label
              className="MuiRating-label"
              htmlFor="hover-feedback-4"
            >
              <span
                className="MuiRating-icon MuiRating-iconEmpty"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </span>
              <span
                className="MuiRating-visuallyhidden"
              >
                4 Stars
              </span>
            </label>
            <input
              checked={false}
              className="MuiRating-visuallyhidden"
              disabled={false}
              id="hover-feedback-4"
              name="hover-feedback"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              type="radio"
              value={4}
            />
            <label
              className="MuiRating-label"
              htmlFor="hover-feedback-5"
            >
              <span
                className="MuiRating-icon MuiRating-iconEmpty"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </span>
              <span
                className="MuiRating-visuallyhidden"
              >
                5 Stars
              </span>
            </label>
            <input
              checked={false}
              className="MuiRating-visuallyhidden"
              disabled={false}
              id="hover-feedback-5"
              name="hover-feedback"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              type="radio"
              value={5}
            />
          </span>
          <h6
            className="MuiTypography-root MuiTypography-subtitle1"
          >
            (4.30) • 10 ratings
          </h6>
        </div>
        <div
          className="favoriteButton"
        >
          <a
            aria-disabled={false}
            className="MuiButtonBase-root MuiButton-root MuiButton-text"
            href="/login"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
          >
            <span
              className="MuiButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeLarge"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
                />
              </svg>
              <span
                className="MuiTypography-root MuiTypography-button"
                style={
                  Object {
                    "color": "white",
                    "marginLeft": "10px",
                  }
                }
              >
                 Add favorite 
              </span>
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </a>
        </div>
        <h5
          className="MuiTypography-root MuiTypography-h5"
          data-testid="overview"
          style={
            Object {
              "marginTop": "20px",
            }
          }
        >
          Overview
        </h5>
        <p
          className="MuiTypography-root MuiTypography-body2"
          style={
            Object {
              "marginTop": "10px",
            }
          }
        >
          Overview
        </p>
      </div>
    </div>
  </div>
</div>
`;
